pkg/api/v1/state.go

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2021 The TestGrid Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package v1 (api/v1) is the first versioned implementation of the API
+package v1
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"net/http"
+	"net/url"
+	"path"
+
+	"github.com/GoogleCloudPlatform/testgrid/config"
+	apipb "github.com/GoogleCloudPlatform/testgrid/pb/api/v1"
+	configpb "github.com/GoogleCloudPlatform/testgrid/pb/config"
+	statepb "github.com/GoogleCloudPlatform/testgrid/pb/state"
+	"github.com/GoogleCloudPlatform/testgrid/util/gcs"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/gorilla/mux"
+)
+
+// findDashboardTab locates dashboard tab in config
+func findDashboardTab(cfg *configpb.Configuration, dashboardKey string, tabKey string) *configpb.DashboardTab {
+	if cfg == nil {
+		return nil
+	}
+	for _, dashboard := range cfg.Dashboards {
+		if config.Normalize(dashboard.Name) == dashboardKey {
+			for _, tab := range dashboard.DashboardTab {
+				if config.Normalize(tab.Name) == tabKey {
+					return tab
+				}
+			}
+			return nil
+		}
+	}
+	return nil
+}
+
+// Grid fetch tab group name grid info (columns, rows, ..etc)
+func (s Server) GroupGrid(configPath *gcs.Path, groupName string) (*statepb.Grid, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), s.Timeout)
+	defer cancel()
+	groupPath, err := configPath.ResolveReference(&url.URL{Path: path.Join(s.GridPathPrefix, groupName)})
+	grid, _, err := gcs.DownloadGrid(ctx, s.Client, *groupPath)
+	if err != nil {
+		return nil, fmt.Errorf("load %s: %v", groupName, err)
+	}
+	return grid, err
+}
+
+// Grid fetch tab and grid info (columns, rows, ..etc)
+func (s Server) Grid(r *http.Request, cfg *configpb.Configuration, dashboardKey string, tabKey string) (*statepb.Grid, string) {
+	tab := findDashboardTab(cfg, dashboardKey, tabKey)
+	if tab == nil {
+		return nil, fmt.Sprintf("Dashboard {%q} or tab {%q} not found", dashboardKey, tabKey)
+	}
+	configPath, _, _ := s.configPath(r)
```

Comment:
If this generated an error, seeing why early would be useful.

Reply:
I skipped this at first because cfg would be nil if s.configPath() would generate an error. And there was a cfg nil check before that call. Now, I agree that if that cfg nil check is removed in the future, this would leave this call exposed to potential exceptions.
Updated.